src/tracer.c

```diff
@@ -1325,26 +1328,28 @@ static MRF_RETURN_TYPE tracing_fn(struct request_queue *q, struct bio *bio)
         smp_rmb();
         tracer_for_each(dev, i)
         {
-                if (!dev || test_bit(UNVERIFIED, &dev->sd_state) ||
-                        !tracer_queue_matches_bio(dev, bio))
-                                continue;
+                if (!tracer_is_bio_for_dev(dev, bio)) continue;
```

Comment:
Hi @jnse 
This fix has a little issues , e.g.:
[Scenario 1]
1. `sda1` and `sdb1`(or even `dm-1` for `md-1`) are in snapshot mode and has the same partition range, 
[Issue]
1.  This will give `true` for `tracer_is_bio_for_dev` check, but failed to `tracer_should_trace_bio`(due to different queue)
[Possible result]
1. `snap_trace_bio(sda1, sda1's bio)`      good
2. `snap_trace_bio(sdb1, sdb1's bio)`      good
3. `SUBMIT_BIO_REAL(sda1, sdb1's bio)`     bad
4. `SUBMIT_BIO_REAL(sdb1, sda1's bio)`     bad

(`sdb1` could  be replaced with `dm-1` or `md1`)

I will keep trying  help think the potential issue scenarios and upload here.

Best Regards,
Nick

Reply:
Should be fixed now with the latest change - I moved the queue check into tracer_is_bio_for_dev ; gendisks are different between physical disks, so the queues should be different as well - 
It does not solve the problem however. The testing I'm doing is on 5.9 and as far as I can tell, the bio's are being submitted to the correct gendisk for the device.